### PR TITLE
Implement Popup dialog to capture iroh-net connection parameters

### LINKED ISSUE #222: Popup dialog to capture iroh-net connection parameters
This is needed in order to complete #106 

To connect to a remote Iroh-net node we need the [Node ID](https://docs.rs/iroh-base/0.19.0/iroh_base/key/type.NodeId.html) . using NodeId::from_str().
e.g. "odvvntniz4qijaq6gdnsxuhe2wlhugiwdgkb7uqfw6zxhke7zxmq"
So, we should have a text entry field of that length for it, with name "nodeid"

An additional field, with name "relay_url", that we can pre-fill before showing the dialog, but that the user can edit is also desired.

A small (x) to clear the text fields would be a nice touch, but not required.

A long text at the top of the dialog, that I will write, to explain what the parameters are for etc is also desired.
If you create it with some placeholder multi-line text, I will re-write the message.

Then a button to connect, or a button to cancel.

When the "Connect" button is pressed, we should disable the button, and the text entry fields while it runs.
we should show a spinner and a "Connecting to remote Pi" text.

A method will need to be called to "COnnect" when the connect button is pressed, that takes the nodeid and relay_url, and that may produce success or a failure.

On Success we should close the dialog.

On failure it should stay open, and display (red text at the bottom?) some error message. Then retry or cancel.

### PR DESCRIPTION
**Pull request by sundaram123krishnan**
Fixes #222

### CODE REVIEW

**Comment by andrewdavidmackenzie** [comment]
For me it takes two ESCAPE keys to close it.
The first seems to remove focus from the first field.
The second seems to close it.
Possible to close it with one ESCAPE key press?

**Comment by andrewdavidmackenzie** [comment]
When I enter some text in both fields, then close dialog with cancel...
When I re-open it, the relay url text is still there and other is blank.

This will probably be fine when we pre-fill that field...just mentioning as I was surprised.

**Comment by andrewdavidmackenzie** [comment]
Can we have the Connect button send a message requesting to connect (with the field strings as data), and NOT close dialog immediately?

(I would add the code to handle that message in update and trigger the subscription to try to connect).

Is it possible for the background connection task to then send messages that:
- show some in-progress connecting status in the dialog (Connecting)
- provide an error on failure that will be shown as red text at the bottom of dialog (ConnectionError)
- close it if connection succeeds (ConnectionSucceeded)

**Comment by andrewdavidmackenzie** [comment]
Could we add some validation on the fields:
- nodeid: We should check length, no whitespace and is a valid Base64 string. I think I can find a Iroh function that will help us check it's a valid nodeid
- relay url: check it's a valid url? You could parse with Url class and check no error

then show the errors in the dialog and do not send the connection message if inputs not valid?

**Comment by andrewdavidmackenzie** [comment]
This class for Iroh NodeId has a verify and from_bytes method:
https://docs.rs/iroh-net/latest/iroh_net/type.NodeId.html

The ideal would be that the Message that "Connect" button sends would send *typed* info
enum Message {
    ConnectIroh(NodeId, Url)
}

That way, invalid values will not be possible in the message handling code.

File: src/piggui.rs [hunk 1]
```diff
@@ -238,7 +355,80 @@ impl Application for Piggui {
             .center_x()
             .center_y();
 
-        self.toast_handler.view(content.into())
+        if self.show_modal {
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:358]: We should move all of this to a view() method of a ConnectDialog (or similar) module...

File: src/piggui.rs [hunk 2]
```diff
@@ -249,6 +439,7 @@ impl Application for Piggui {
     fn subscription(&self) -> Subscription<Message> {
         let subscriptions = vec![
             iced::event::listen().map(WindowEvent),
+            event::listen().map(ModalKeyEvent),
```
Comment (andrewdavidmackenzie) [inline src/piggui.rs:442]: We only need to add this subscription while the dialog is open, no?
If so, I would do like I do elsewhere, which is makde subscriptions a mut vec.... then add it only if show_modal == true.

File: src/views/hardware_view.rs [hunk 3]
```diff
@@ -574,14 +575,16 @@ fn filter_options(
         .cloned()
         .collect();
 
-    if !config_options.contains(&PinFunction::None) && selected_function.is_some() && selected_function != Some(PinFunction::None) {
+    if !config_options.contains(&PinFunction::None)
+        && selected_function.is_some()
+        && selected_function != Some(PinFunction::None)
+    {
```
Comment (andrewdavidmackenzie) [inline src/views/hardware_view.rs:581]: just formatting changes, right?

Has something changed regarding formatting settings again?
Was it me?

File: src/views/hardware_view.rs [hunk 4]
```diff
@@ -703,7 +705,10 @@ mod test {
 
         // Test case: No function selected
         let result = filter_options(&options, None);
-        assert_eq!(result, vec![PinFunction::Input(None), PinFunction::Output(None)]);
+        assert_eq!(
+            result,
+            vec![PinFunction::Input(None), PinFunction::Output(None)]
+        );
```
Comment (andrewdavidmackenzie) [inline src/views/hardware_view.rs:711]: these are all new formatting change, right?

**Review by andrewdavidmackenzie — CHANGES_REQUESTED** [review]
I think I would strive to do the refactoring now, before merge, and keep changes to piggui.rs minimal, and capsulate as much as possible in a ConnectDialog struct and module...

File: src/piggui.rs [hunk 5]
```diff
@@ -149,6 +179,93 @@ impl Application for Piggui {
...
+                                // Make iroh connection
+                                // Add spinner when establishing remote connection
+                            }
+                            Err(err) => {
+                                self.iroh_connection_error = format!("{}", err);
+                            }
+                        }
+                    }
+                    Err(err) => {
+                        self.iroh_connection_error = format!("{}", err);
```
[the remark this answers is not in the document]
Comment (sundaram123krishnan) [reply src/piggui.rs:209]: Yes

File: src/views/hardware_view.rs (hunk 3 above)
Comment (sundaram123krishnan) [reply src/views/hardware_view.rs:581]: I didn't commit any changes in hardware_view

File: src/views/hardware_view.rs (hunk 4 above)
Comment (sundaram123krishnan) [reply src/views/hardware_view.rs:711]: not sure, i haven't staged this changes

**Review by andrewdavidmackenzie — APPROVED** [review]

**Pull request merged**